kore/src/Kore/Unification/Substitution.hs

```diff
@@ -35,71 +38,108 @@ import           Data.Set
 import qualified Data.Set as Set
 import           GHC.Generics
                  ( Generic )
+import           GHC.Stack
+                 ( HasCallStack )
 import           Prelude hiding
                  ( null )
 
 import Kore.Step.Pattern as Pattern
 import Kore.TopBottom
        ( TopBottom (..) )
 
--- | 'Substitution' is a wrapper for a list of substitutions of the form
--- (variable level, StepPattern level variable). Values of this type should be
--- manipulated using the functions in this module.
+{- | @Substitution@ represents a collection @[xᵢ=φᵢ]@ of substitutions.
+
+Individual substitutions are a pair of type
+
+@
+(variable level, StepPattern level variable)
+@
+
+A collection of substitutions @[xᵢ=φᵢ]@ is /normalized/ if, for all @xⱼ=φⱼ@ in
+the collection, @xⱼ@ is unique among all @xᵢ@ and none of the @xᵢ@ (including
+@xⱼ@) occur free in @φⱼ@.
+
+ -}
 data Substitution level variable
+    -- TODO (thomas.tuegel): Instead of a sum type, use a product containing the
+    -- normalized and denormalized parts of the substitution together. That
+    -- would enable us to keep more substitutions normalized in the Semigroup
+    -- instance below.
```

Comment:
Are you sure it's worth doing that? What would be the benefit? And it might not be trivial - consider `x=f(y)`: it's ok to merge it with `z=g(y)`, but not with any of `x=g(z)`, `y=g(z)`, `z=g(x)`, or `z=sigma(y) /\ z = sigma(g(t))`.

Reply:
I think it's worth it if we keep substitution deduplication as a separate step, because we can keep the substitution normalized (well, semi-normalized: cycle-free) incrementally. The most expensive thing we do during execution is merging and re-normalizing substitutions; it's expensive in large part because merging/re-normalization always begins by assuming nothing is normalized. I have in mind a representation such as `data Substitution = Substitution { normal = Map variable (MultiAnd pattern), denormal = Map variable (MultiAnd pattern) }` with the invariant that no variable occurs as a key in both `normal` and `denormal` and the keys of `normal` do not occur in any values of `normal` or `denormal`.

I would merge those examples in the following ways, where each substitution is represented as a pair as described above:

1. `([x=f(y)], []) ∧ ([x=g(z)], [])` = `([x=f(y)∧g(z)], [])` (postponing de-duplication)
1. `([x=f(y)], []) ∧ ([y=g(z)], [])` = `([x=f(g(z)), y=g(z)], [])` (applying substitution, which has no cycles)
1. `([x=f(y)], []) ∧ ([z=g(x)], [])` = `([x=f(y), z=g(f(y))], [])` (as above)
1. `([z=σ(y)], []) ∧ ([z=σ(g(t))], [])` = ([z=σ(y)∧σ(g(t))], [])` (postponing de-duplication, as above)
1. An interesting case not mentioned: `([x=f(y), []) ∧ ([y=g(x)], [])` = `([x=f(y), y=g(f(y))])` (breaking a cycle by applying the `normal` substitution and denormalizing; there is an choice about which way we break the cycle)

Anyway, we can discuss this in more detail when performance becomes an issue, and we can decide to if and how to do this, or not. For now, I mainly wanted to document that we have the option of doing this.